### PR TITLE
Improve performance of Hamming computer

### PR DESCRIPTION
**Pull request by shengjun1985**
Signed-off-by: shengjun.li <shengjun.li@zilliz.com>

Improve performance of Hamming computer

### CODE REVIEW

File: faiss/utils/hamming-inl.h [hunk 1]
```diff
@@ -226,19 +227,52 @@ struct HammingComputerDefault {
...
+        quotient8 = code_size / 8;
+        remainder8 = code_size % 8;
     }
 
     int hamming (const uint8_t *b8) const {
         int accu = 0;
-        for (int i = 0; i < n; i++)
-            accu += popcount64 (a[i] ^ b8[i]);
+
+        auto a64 = reinterpret_cast<const uint64_t *>(a8);
```
Comment (mdouze) [inline faiss/utils/hamming-inl.h:238]: please use explicit types (no auto)

File: faiss/utils/hamming-inl.h [hunk 2]
```diff
@@ -226,19 +227,52 @@ struct HammingComputerDefault {
...
+                    case 3: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 2: accu += popcount64(a64[i] ^ b64[i]); i++;
+                    case 1: accu += popcount64(a64[i] ^ b64[i]); i++;
+                }
+        }
+        if (remainder8) {
+            auto a = a8 + 8 * quotient8;
+            auto b = b8 + 8 * quotient8;
+            switch (remainder8) {
+                case 7: accu += hamdis_tab_ham_bytes[a[6] ^ b[6]];
```
Comment (mdouze) [inline faiss/utils/hamming-inl.h:259]: I suspect the table look-up is slower than using popcount64

File: faiss/utils/hamming-inl.h [hunk 3]
```diff
@@ -226,19 +227,52 @@ struct HammingComputerDefault {
...
     int hamming (const uint8_t *b8) const {
         int accu = 0;
-        for (int i = 0; i < n; i++)
-            accu += popcount64 (a[i] ^ b8[i]);
+
+        auto a64 = reinterpret_cast<const uint64_t *>(a8);
+        auto b64 = reinterpret_cast<const uint64_t *>(b8);
+        int i = 0, len = quotient8;
+        switch (len & 7) {
+            default:
```
Comment (mdouze) [inline faiss/utils/hamming-inl.h:242]: I did not know that jumping there from a switch was possible...

**Review by mdouze — COMMENTED** [review]
Looks good. 
Do you have some evidence that it is indeed faster? 
Please add a small timing script in benchs/ so that we keep a trace of what benchmark was used to take that decision.

File: faiss/utils/hamming-inl.h (hunk 1 above)
Comment (shengjun1985) [reply faiss/utils/hamming-inl.h]: Done.

File: faiss/utils/hamming-inl.h (hunk 2 above)
Comment (shengjun1985) [reply faiss/utils/hamming-inl.h:259]: Here we calculate 1 to 7 uint8.
And my local tests show that table look-up is much faster than using popcount64.

benchmark with code_size=127 n=4000000 nrun=100
Hamming_Dft_lookup  implem: 550.326 ms
Hamming_Dft_popcnt  implem: 743.236 ms
benchmark with code_size=255 n=4000000 nrun=100
Hamming_Dft_lookup  implem: 1117.139 ms
Hamming_Dft_popcnt  implem: 1262.516 ms

**Comment by shengjun1985** [comment]
[My local machine]
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz * 8
Memory:  15.5 GiB

[Output of "bench_hamming_computer.cpp" on my local machine]
benchmark with code_size=128 n=4000000 nrun=100
Hamming_Dft  implem: 533.907 ms
Hamming_M8   implem: 570.451 ms
Hamming_M4   implem: 1111.180 ms
benchmark with code_size=256 n=4000000 nrun=100
Hamming_Dft  implem: 1058.063 ms
Hamming_M8   implem: 1140.435 ms
Hamming_M4   implem: 2159.075 ms
benchmark with code_size=512 n=4000000 nrun=100
Hamming_Dft  implem: 2149.452 ms
Hamming_M8   implem: 2253.472 ms
Hamming_M4   implem: 4304.233 ms
benchmark with code_size=1000 n=4000000 nrun=100
Hamming_Dft  implem: 4138.793 ms
Hamming_M8   implem: 4370.501 ms
Hamming_M4   implem: 8341.167 ms

File: faiss/utils/hamming-inl.h (hunk 3 above)
Comment (shengjun1985) [reply faiss/utils/hamming-inl.h:242]: I refer to the [Duff's device](https://en.wikipedia.org/wiki/Duff%27s_device) here.

**Comment by mdouze** [comment]
Looks good to me, importing to validate on the FB internal version.

**Pull request closed without merging**